pageserver/src/tenant.rs

```diff
@@ -952,11 +1019,95 @@ impl Tenant {
         Ok(totals)
     }
 
-    pub fn tenant_id(&self) -> TenantId {
-        self.tenant_id
+    /// - run initdb to init temporary instance and get bootstrap data
+    /// - after initialization complete, remove the temp dir.
+    async fn bootstrap_timeline(
+        &self,
+        conf: &'static PageServerConf,
+        timeline_id: TimelineId,
+    ) -> Result<Arc<Timeline>> {
+        // XXX: ensure no other condurrent processes run initdb for the new timeline
+        let timelines_guard = self.timelines.lock();
```

Comment:
Instead of grabbing the lock, I'd suggest running initdb in a known unique path. So instead of `.temp` suffix, something like `.<random string>.temp`. That way, there cannot be a conflict while running initdb.

Reply:
That is a bit complicated, because it hardens the tmp dir check on restart: we would need to somehow detect such directories for the timeline and then to somehow verify that the timeline was initialized.

With the PR proposed model, if we find such directory near the timeline, we know that the timeline was not initialized correctly (due to that tmp dir being last) hence both tmp dir and the timeline need purging.

Yet with multiple such directories, we can import initdb result from one of the runs and fail the others with pageserver restart so, that there are some tmp initdb dirs and a properly initialized timeline.
I would imagine, it's ok to check `disk_consistent_lsn` in such case to ensure it's on 0 and it should be enough, but it's rather subtle.

Reply:
> With the PR proposed model, if we find such directory near the timeline, we know that the timeline was not initialized correctly (due to that tmp dir being last) hence both tmp dir and the timeline need purging.

Oh, that's very subtle! So if there are two directories at pageserver startup:
```
tenants/1234/timelines/5678.temp
tenants/1234/timelines/5678
```
It will remove *both*?

Ok, yes, that works. Needs a big comment somewhere! And makes me a bit uncomfortable; it seems safe to delete anything with the `.temp` suffix, but it feels much more dangerous to delete something that looks valid, just because a different directory with `.temp` suffix exists.

IMHO a nicer way to do this would be to create the directory for the new, empty, timeline with `.temp` suffix, import all the data there, and finally once it's ready, rename it without the .temp suffix. That would require more code changes though. Currently we assume in many places that the path for a timeline is always `tenants/<tenant_id>/timelines/<timeline_id>`, and now during import it would be `tenants/<tenant_id>/timelines/<timeline_id>.temp`.

In summary, I'm OK with this PR for now, but needs a comment somewhere, and in the long run I think we should consider that other approach.

Reply:
> It will remove both?

Not right now, I've planned to add that in another PR, but will add here, since apparently very related.
But yes, that's the intention and that's how we're sometimes left with 0-init "broken" timelines after load tests, presumably: I see no other way but to remove both directories with the current code approach.

@LizardWizzard and I discussed tombstones a bit as a solution to the issue, but it all looks quite tangled now, so whatever we try, it will require plenty of changes.
Not happy with lock acquisition either, but see no other viable options.

I wonder, how appropriate it is for such operations to go and create the tmp directory right away, without any locks?
If all concurrent operations fail (aborting tenant/timeline/whatever creation) and only one creates the directory and proceeds with init, we get ourselves a "good" race without extra locking?